submissions/OlexiyDobroskok/OOP_exercise/index.js

```diff
@@ -0,0 +1,159 @@
+  friends: "",
+};
+
+const humanDefaultProperties = {
+  hands: 2,
+  legs: 2,
+};
+
+const petsDefaultProperties = {
+  paws: 4,
+};
+
+class Inhabitant {
+  static worldPopulation = [];
+  constructor({
+    species,
+    name,
+    gender,
+    saying,
+    friends,
+  } = inhabitantDefaultProperties) {
+    this.species = species ?? inhabitantDefaultProperties.species;
+    this.name = name ?? inhabitantDefaultProperties.name;
+    this.gender = gender ?? inhabitantDefaultProperties.gender;
+    this.saying = saying ?? inhabitantDefaultProperties.saying;
+    this.friends = friends ?? inhabitantDefaultProperties.friends;
+    Inhabitant.worldPopulation.push(this);
+  }
+
+  addFriend({ name }) {
+    this.friends += name + "; ";
+  }
+
+  introduceYourSelf() {
+    if (this.friends === "") {
+      this.friends = "Looking for friends!";
+    }
+    return `<strong>${this.saying}! My name is ${this.name}. Species: ${this.species}. Gender: ${this.gender}. Friends: ${this.friends} </strong>`;
+  }
+}
+
+class Human extends Inhabitant {
+  constructor({
+    name,
+    gender,
+    saying,
+    legs,
+    hands,
+    friends,
+  } = humanDefaultProperties) {
+    super({ species: "human", name, gender, saying, friends });
+    this.hands = hands ?? humanDefaultProperties.hands;
+    this.legs = legs ?? humanDefaultProperties.legs;
+  }
+  introduceYourSelf() {
+    return (
+      super.introduceYourSelf() +
+      `<strong>I have ${this.hands} hands and ${this.legs} legs!</strong>`
+    );
+  }
+}
+
+class Man extends Human {
+  constructor({ name, saying, friends, legs, hands }) {
+    super({ name, gender: "male", saying, legs, hands, friends });
+  }
+}
+
+class Woman extends Human {
+  constructor({ name, saying, friends, legs, hands }) {
+    super({ name, gender: "female", saying, legs, hands, friends });
+  }
+}
+
+class Pets extends Inhabitant {
+  constructor({
+    species,
+    name,
+    gender,
+    paws,
+    saying,
+    friends,
+  } = petsDefaultProperties) {
+    super({ species, name, gender, saying, friends });
+    this.paws = paws ?? petsDefaultProperties.paws;
+  }
+  introduceYourSelf() {
+    return (
+      super.introduceYourSelf() + `<strong>I have ${this.paws} paws!</strong>`
+    );
+  }
+}
+
+class Cat extends Pets {
+  static meow;
+  constructor({ name, gender, saying = "¡miau miau", friends, paws }) {
+    super({
+      species: "cat",
+      name,
+      gender,
+      paws,
+      saying,
+      friends,
+    });
+    Cat.meow = this.saying;
+  }
+}
+
+class Dog extends Pets {
+  static bark;
+  constructor({ name, gender, saying = "¡guau guau", friends, paws }) {
+    super({
+      species: "dog",
+      name,
+      gender,
+      paws,
+      saying,
+      friends,
+    });
+    Dog.bark = this.saying;
+  }
+}
+
+class CatWoman extends Woman {
+  constructor({ name, saying = "", friends, legs, hands }) {
+    super({ name, saying, friends, legs, hands });
+    this.saying = Cat.meow + " " + saying;
```

Comment:
Use static property is a good start in the right direction.
But here is the problem: if we change cat's defaut meow (which is allowed by Cat class API), cat-women's saying will be out of sync. Can you devise a way to bind their saying more tightly? You are only one step away from the solution.

Reply:
@OleksiyRudenko The easiest way I see is to assign a default value to the static property of the cat:
Cat.meow = "¡miau miau";
    this.saying = Cat.meow;
but then all cats lose their identity and it makes me sad =)))

Reply:
As long as property holds a primitive value the tie is weak, i.e. simply copied once. Strings, numbers and booleans are primitives. We need something that is passed by reference. These are JS objects (arrays and functions are also objects). Class methods can also be static. There are also very specific methods - getters and setters.
Given the above find a solution that will help to both meet requirements and minimize impact on code to keep it readable (i.e. we do not want dogs and humans to have added complexity just becase we need to resolve a matter between cats and cat-women).

Reply:
@OleksiyRudenko Tried to bind Woman and cat by static method. Looks strange)

Reply:
It does. Truth is that hybrids do not fit OOP hierarchical nature really well.
Composition is a better pattern in similar cases.
The purpose of this part of task was exactly to see the weirdness of how it has to be implemented in OOP paradigm.

Reply:
@OleksiyRudenko Good task. At a glance nothing difficult, but on the small details forces to revise a lot of materials). Do I still have to make some adjustments to the task?